### PR TITLE
feat(hub): adding snapshot download method

### PR DESCRIPTION
**Pull request by axel7083**
## Description

We can now create a `snapshotDownload` method similator to the `snapshot_download` of the PY lib[^1], clone to the cache (only cache supported for now) a repository (either model, space or dataset)

[^1]: https://huggingface.co/docs/huggingface_hub/en/guides/download#download-an-entire-repository

## Related issues/PR

With the amazing help of @coyotte508 we were able to merge the following changes

- https://github.com/huggingface/huggingface.js/pull/1034
- https://github.com/huggingface/huggingface.js/pull/1031
- https://github.com/huggingface/huggingface.js/pull/999

Which allow this PR to provide a python compliant clone of a hugging face repository to the cache directory. 

## Testing

- [x] unit tests are covering the new feature

**Manually**

```ts
await snapshotDownload({
	repo: {
		name: 'OuteAI/OuteTTS-0.1-350M',
		type: 'model',
	},
});
```

assert using the `huggingface-cli` tool (python)
```
$: huggingface-cli scan-cache
REPO ID                             REPO TYPE SIZE ON DISK NB FILES LAST_ACCESSED     LAST_MODIFIED     REFS LOCAL PATH                                                                         
----------------------------------- --------- ------------ -------- ----------------- ----------------- ---- ---------------------------------------------------------------------------------- 
OuteAI/OuteTTS-0.1-350M             model           731.6M       14 5 minutes ago     5 minutes ago     main /home/axel7083/.cache/huggingface/hub/models--OuteAI--OuteTTS-0.1-350M
```

### CODE REVIEW

File: packages/hub/src/lib/snapshot-download.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+import { datasetInfo } from "./dataset-info";
+import { modelInfo } from "./model-info";
+import { toRepoId } from "../utils/toRepoId";
+import { join, dirname } from "node:path";
+import { mkdir, writeFile } from "node:fs/promises";
+import { downloadFileToCacheDir } from "./download-file-to-cache-dir";
+
+export const DEFAULT_REVISION = "main";
+
+export async function snapshotDownload(
```
Comment (coyotte508) [inline packages/hub/src/lib/snapshot-download.ts:14]: Can you add a doc comment - saying it downloads a revision of a repo to local file system? And that it returns the local path. And that there are related methods (`downloadFileToCacheDir`, and `scanCachedRepo`) ?

(thinking about it, we could also do web storage with [indexed db](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Client-side_web_APIs/Client-side_storage#storing_complex_data_%E2%80%94_indexeddb) - but let's do that if ever in another PR :sweat: )

**Review by coyotte508 — APPROVED** [review]
Nice! Thanks for the full feature

File: packages/hub/src/lib/snapshot-download.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+import { mkdir, writeFile } from "node:fs/promises";
+import { downloadFileToCacheDir } from "./download-file-to-cache-dir";
+
+export const DEFAULT_REVISION = "main";
+
+/**
+ * Downloads an entire repository at a given revision in the cache directory {@link getHFHubCachePath}.
+ * You can list all cached repositories using {@link scanCachedRepo}
+ * @remarks It uses internally {@link downloadFileToCacheDir}.
+ * @param params
```
Comment (coyotte508) [inline packages/hub/src/lib/snapshot-download.ts:18]:
```suggestion
```

**Review by coyotte508 — COMMENTED** [review]
And can you add a small section in `packages/hub/README.md` regarding cache management? :pray:

**Comment by axel7083** [comment]
> And can you add a small section in `packages/hub/README.md` regarding cache management? 🙏

Yes np!

File: packages/hub/README.md [hunk 3]
```diff
@@ -117,14 +117,51 @@ Checkout the demo: https://huggingface.co/spaces/huggingfacejs/client-side-oauth
 
 The `@huggingface/hub` package provide basic capabilities to scan the cache directory. Learn more about [Manage huggingface_hub cache-system](https://huggingface.co/docs/huggingface_hub/en/guides/manage-cache).
 
+### `scanCacheDir`
+
+You can get the list of cached repository using the `scanCacheDir` function.
```
Comment (coyotte508) [inline packages/hub/README.md:122]:
```suggestion
You can get the list of cached repositories using the `scanCacheDir` function.
```

**Comment by axel7083** [comment]
> I just reviewed them and logic is good. I'm a bit worried though about the lack of locks when downloading an individual file. In `huggingface_hub`, we use [`filelock`](https://pypi.org/project/filelock/) which guarantees that a single file is never downloaded twice concurrently. In the js implementation, since the `(...).incomplete` filename is deterministic, it can happen that 2 processes are downloading bytes to the same file concurrently (and in this case I expect one process to crash?). Is there an equivalent that could be used in JS to avoid concurrency issues?

This is a very complicated problem actually, I've just spent an hour investigating.

The python library [`filelock`](https://pypi.org/project/filelock/) use a very complex mechanism to lock the file in a way, that other processes cannot delete the file. 

The problem is that the python library apply _a real lock_ on windows it uses [msvcrt](https://docs.python.org/3/library/msvcrt.html) which provide advanced File Operations and on unix system [fcntl](https://docs.python.org/3/library/fcntl.html)

There are using platform dependent mechanism to provide a real lock.

## JS packages

There is currently two well known packages:

- https://github.com/npm/lockfile seems to only create the file, and check for existence. 
- https://www.npmjs.com/package/proper-lockfile uses the `mkdir` method to lock a file.

**Comment by coyotte508** [comment]
If we open a file in write mode, no two processes can open it, so it suffices as a lock?

In any case, let's not add additional deps

**Comment by axel7083** [comment]
> If we open a file in write mode, no two processes can open it at the same time, so it suffices as a lock?

No this is not the case in Node sadly, you can check it by deleting the <path> file through the windows explorer.

```ts
const file = await open(path, 'w');

setTimeout(async () => {
    await file.close();
}, 60_000);
```

**Comment by Wauplin** [comment]
> no two processes can open it at the same time

What happens if we try to do so? An error is triggered? If that's the case, we could catch the error, wait while the file is been downloaded by the other process and then continue without downloaded once the first process has completed.

I'm raising the question mainly because concurrency issues caused quite some trouble at some point on the Python side. When using multiple GPUs with 1 GPU == 1 process, each process tries to download the weight file at the same time - hence causing the concurrency issue. I guess this is much less likely to happen in JS but still want to raise the question.

**Comment by axel7083** [comment]
> What happens if we try to do so? An error is triggered? 

No error are triggered sadly, we may check for file existence, but the python implementation is complex and change the file descriptor using C code to interact with os level structure.

We could from the JS detect if the file is locked, so the JS would avoid downloading the file if the python is doing it, but we cannot prevent the python to do it if the JS is downloading it :/

**Comment by coyotte508** [comment]
Note that we can open a file in Node.JS with mode "wx", so it fails if the file already exists.

**Comment by axel7083** [comment]
> Note that we can open a file in Node.JS with mode "wx", so it fails if the file already exists.

Yes, but we can stat to check if it exists, that not the issue, the problem is telling the python that we lock the file, which we cannot :(

**Comment by coyotte508** [comment]
well @Wauplin is the maintainer of the python lib :)

**Comment by Wauplin** [comment]
Ah, I wasn't thinking of a lock shared between JS and Python (that seems much less likely to happen). I think handling the concurrency issue between JS processes is already good enough

**Comment by axel7083** [comment]
> Ah, I wasn't thinking of a lock shared between JS and Python (that seems much less likely to happen). I think handling the concurrency issue between JS processes is already good enough

Oh yeah, if that's the only thing we can definitely use https://github.com/npm/lockfile

**Comment by coyotte508** [comment]
We can copy the mechanism internally (or just use `wx` write mode when downloading files) - but let's not add dependencies

**Pull request merged**